### ISSUE TITLE
No data in user reports

### ISSUE THREAD

**Issue by ehuggett** [iomad/iomad#321]
No data was shown in the user report for users which have enrolled into courses in 3.0

I think "noresults" should be false if there are results, the enrolled courses are shown after making this change.

```
diff --git a/local/report_users/userdisplay.php b/local/report_users/userdisplay.php
index 6f2a64d..6184cbc 100644
--- a/local/report_users/userdisplay.php
+++ b/local/report_users/userdisplay.php
@@ -171,7 +171,7 @@ $noresults = true;

 foreach ($usercourses as $usercourse) {
     if ($usercompletion[$usercourse->id] = userrep::get_completion($userid, $usercourse->id, $showhistoric) ) {
-        $noresults = true;
+        $noresults = false;
         $usercourseid = $usercourse->id;

         // Check if the course is also in progress.
```

However (separate bug), clicking "(user details)" on any of the results after applying the patch above causes what looks like a var_dump of a php array to be displayed. I will look into this tomorrow and hopefully be able to submit a patch before the end of the day

**Comment by turf212** [comment]
From memory the userrep::get_completion never returns a false which is why the noresults check was put in another bit of logic.  I also need to have a look at that and see where the logic is failing.

What were your steps to get the issue?

**Comment by ehuggett** [comment]
To reproduce the first, open the user report for any user who has enrolled into one or more courses and "no results" will be displayed when it should list the enrolled courses.

For the second issue click the "(user details)" link in one of the course rows (after fixing the first issue). I found the cause, its a deliberate print_r... possibly something left over from debugging? https://github.com/iomad/iomad/blob/IOMAD_30_STABLE/local/report_users/userdisplay.php#L323-L325

Patch for both issues (I've also renamed the variable to avoid the double negative "!$noresults")

```
diff --git a/local/report_users/userdisplay.php b/local/report_users/userdisplay.php
index 6f2a64d..97e4ec3 100644
--- a/local/report_users/userdisplay.php
+++ b/local/report_users/userdisplay.php
@@ -167,11 +167,11 @@ $compusertable->head[] = get_string('actions', 'local_report_users');
 $compusertable->align[] = 'center';

 // Set that there is nothing found here first.
-$noresults = true;
+$results = false;

 foreach ($usercourses as $usercourse) {
     if ($usercompletion[$usercourse->id] = userrep::get_completion($userid, $usercourse->id, $showhistoric) ) {
-        $noresults = true;
+        $results = true;
         $usercourseid = $usercourse->id;

         // Check if the course is also in progress.
@@ -310,7 +310,7 @@ if (empty($dodownload)) {
     }

     // If we have anything show it.
-    if (!$noresults) {
+    if ($results) {
         echo html_writer::table($compusertable);
     } else {
         echo "</br><b>" . get_string('noresults') . "</b>";
@@ -319,9 +319,6 @@ if (empty($dodownload)) {

 if (!empty($courseid && !empty($usercompletion[$courseid]))) {
     if (empty($dodownload)) {
-        echo "usercompletion = <pre>";
-        print_r($usercompletion);
-        echo "</pre></br>";
         echo "<h3>".$usercompletion[$courseid]->data[$courseid]->coursename.
              " (<a href='".
              new moodle_url('/local/report_completion/index.php', array('courseid' => $courseid)).
```

**Comment by turf212** [comment]
Confirmed and I'll push your fix ASAP. Thanks :)

**Comment by turf212** [comment]
So there is still an issue with this fix (which is why I havent pushed it).  Steps:
create a couple of courses which have completion info for them
Enrol a user onto the courses
wait for cron to catch up with the completion
User report for that user shows no data
Course report summary shows three people enrolled but not started
Click on course detail, nothing is displayed

It should be showing enrolled and not started users as well as users who are started and users who are completed.  Started is the Moodle definition of started.  I.e. they have completed one of the course completion criteria.  If you only have one in the course the user goes straight from not started to started.

I need to look at this more fully. :/

**Comment by turf212** [comment]
Ok the issue I was having is with the SQL to populate the temporary table.  The join to the grade items table is an INNER join which means nothing is being populated if there are no grade items in the course.  I've changed this to a LEFT JOIN (which is what it already was on the grade_grades table) for both the user completion and the course completion reports and this has fixed it.  I've also tidied up the course completion report logic as it wasn't correct for started and also there was a missing string for the certificate link if there were no certificates in the system.  This appears to be the same issue as #270
